### DISCUSSION TITLE
neg_ratio variable object detection error

### DISCUSSION THREAD

**Discussion by GisKamal** [azavea/raster-vision#2021] [Q&A] [answered]
Hi,

I am trying to train an object detection model using the most recent master branch of raster vision , and running with a docker image on ubuntu machine i get the following error:
 
 
raise ValueError(
ValueError: neg_ratio specified, but no bboxes found in scene.

the following part of the script that uses the neg_ratio variable is : 
 
 
    chip_options = ObjectDetectionChipOptions(neg_ratio=0.95,ioa_thresh=0.75)

    if nochip:
        window_opts = ObjectDetectionGeoDataWindowConfig(
            method=GeoDataWindowMethod.random,
            size=chip_sz,
            size_lims=(chip_sz, chip_sz + 1),
            max_windows=500,
            max_sample_attempts=100,
            clip=True,
            neg_ratio=chip_options.neg_ratio,
            ioa_thresh=chip_options.ioa_thresh,
            neg_ioa_thresh=0.2)
            
            
any suggestions or corrections ??

**Comment by AdeelH** [comment]
There are either no bboxes in the scene or the labels and imagery are misaligned. For the latter, make sure that the labels GeoJSON is in epsg:4326 coordinates and that it lines up correctly with the imagery when opened in QGIS.

**Comment by GisKamal** [comment]
Hi, Thanks for your reply, I have checked my GeoJSON labels using QGIS, and they line up perfectly with my tif image, concerning the CRS it's EPSG:4326 for both images and labels. One important think to mention that I hope it could help, its that after running my python file I get this message :

2024-01-09 11:40:57:rastervision.pytorch_learner.learner: INFO - Building datasets ...
2024-01-09 11:40:57:rastervision.core.data.vector_source.geojson_vector_source: INFO - Ignoring CRS ({'type': 'name', 'properties': {'name': 'urn:ogc:def:crs:OGC:1.3:CRS84'}}) specified in /opt/src/metadata/Images_Labels/1.geojson and assuming EPSG:4326 instead.

any idea please on how to resolve the neg_ratio problem ?

**Reply by AdeelH** [reply]
Like that warning suggests, you might need to transform the coordinates in your GeoJSON file to EPSG:4326 (i.e. lon, lat).

**Reply by GisKamal** [reply]
In layer properties from QGIS, the CRS of my GeoJSON layer is EPSG:4326. One important thing is that using method=GeoDataWindowMethod.sliding and neg_ratio=chip_options.neg_ratio in ObjectDetectionGeoDataWindowConfig arguments, all works fine with no error.

**Reply by AdeelH** [reply]
Can you share how you're defining the label source config and also a sample label file?

**Reply by GisKamal** [reply]
Hi, 
Here the code I'm using to define the raster and vector sources :
```py
def make_scene(id: str) -> SceneConfig:
        raster_uri = "{}{}.tif".format(path_tifs,id)
        label_uri = "{}{}.geojson".format(path_labels,id)

        if test:
            crop_uri = join(processed_uri, 'crops',
                            os.path.basename(raster_uri))
            save_image_crop(
                raster_uri,
                crop_uri,
                label_uri=label_uri,
                size=2000,
                min_features=5)
            raster_uri = crop_uri

        raster_source = RasterioSourceConfig(
            uris=[raster_uri], channel_order=[0, 1, 2])

        vector_source = GeoJSONVectorSourceConfig(
            uris=label_uri, ignore_crs_field=True)
        label_source = ObjectDetectionLabelSourceConfig(
            vector_source=vector_source)

        return SceneConfig(
            id=id, raster_source=raster_source, label_source=label_source)

    class_config = ClassConfig(names=['vehicle'])
    scene_dataset = DatasetConfig(
        class_config=class_config,
        train_scenes=[make_scene(id) for id in train_ids],
        validation_scenes=[make_scene(id) for id in val_ids])
```
And here a sample label data :
[sample_label.zip](https://github.com/azavea/raster-vision/files/13887113/sample_label.zip)

**Reply by AdeelH** [reply, marked as answer]
You need to also specify a `ClassInferenceTransformerConfig` to map polygons in your GeoJSON to class IDs as shown here e.g.: https://github.com/azavea/raster-vision/blob/fe1d152a1b0cdc95aaaf89193b372a35bcd487a0/rastervision_pytorch_backend/rastervision/pytorch_backend/examples/object_detection/cowc_potsdam.py#L92-L97

If you have just one class, the above example should work.
